api/src/main/java/org/apache/iceberg/types/PruneColumns.java

```diff
@@ -77,43 +88,48 @@ public Type struct(Types.StructType struct, List<Type> fieldResults) {
   @Override
   public Type field(Types.NestedField field, Type fieldResult) {
     if (selected.contains(field.fieldId())) {
-      return field.type();
+      if (field.type().isStructType() && !selectFullTypes) {
```

Comment:
Similar to below, it is probably going to be more readable later to handle `selectFullTypes` first and then move to other cases.